### PR TITLE
Upgrade infrastructure

### PR DESCRIPTION
**Pull request by dalonsoa**
Among the many things, this PR

- [x] Uses `pyproject.toml` for project management
- [x] Upgrades docker image version
- [x] Upgrades versions of other services (eg. Azurite, etc.)
- [x] Uses our current setup for GitHub workflows
- [x] Updates code to use modern dependencies way of doing things (eg. bootstrap)
- [x] Replaces `isort`, `flake8` and `black` with `ruff`
- [x] Update pre-commit to use our current list of hooks

Close https://github.com/ImperialCollegeLondon/rse_maintenance_tasks/issues/10

### CODE REVIEW

**Comment by dalonsoa** [comment]
The vast majority of what appear as changed is the linters and formatters.

File: Dockerfile [hunk 1]
```diff
@@ -1,6 +1,4 @@
-FROM python:3.8-slim-buster as python
-
-FROM python
+FROM python:3.12-bullseye
```
Comment (AdrianDAlessandro) [inline Dockerfile]: We probably want this to be a `slim` package. In DUNE we used `python:3.12-slim-bookworm`
```suggestion
FROM python:3.12-slim-bookworm
```

File: docs/README.md [hunk 2]
```diff
@@ -6,8 +6,6 @@ Liionsden is a Web application comprising a database of electrochemical (Lithium
 Ideas drawn from:
 
 * 'Galvanalyser' by Luke Pitt and Prof. David Howey
-* [liionsden by Ivan Kotegov at Imperial College](https://github.com/FaradayInstitution/liionsden)
 * [Universal Battery Database by Samuel Buteau](https://github.com/Samuel-Buteau/universal-battery-database)
-* [dfndb](https://github.com/ndrewwang/dfndb) by @ndrewwang
```
Comment (AdrianDAlessandro) [inline docs/README.md:11]: Why are these removed?

File: management/apps.py [hunk 3]
```diff
@@ -6,4 +6,4 @@ class ManagementConfig(AppConfig):
     name = "management"
 
     def ready(self):
-        import management.signals.handlers
+        pass
```
Comment (AdrianDAlessandro) [inline management/apps.py]: This should be necessary if these  signal handlers are still required

**Review by AdrianDAlessandro — COMMENTED** [review]
Haven't had the chance to actually test it. It all seems good to me. I've made a couple of observations

File: management/apps.py (hunk 3 above)
Comment (dalonsoa) [reply management/apps.py]: Indeed! Well spotted

File: docs/README.md (hunk 2 above)
Comment (dalonsoa) [reply docs/README.md:11]: That repo does not seem to exist, but then I thought it might be private. I'll leave it there, indicate that is private and tell the link checker not to check it.

**Comment by AdrianDAlessandro** [comment]
When I tried running this I got this:
```
Attaching to faraday-liionsden-azurite-1, faraday-liionsden-db-1, faraday-liionsden-web-1
faraday-liionsden-db-1       | 
faraday-liionsden-db-1       | PostgreSQL Database directory appears to contain a database; Skipping initialization
faraday-liionsden-db-1       | 
faraday-liionsden-azurite-1  | Azurite Blob service is starting at http://0.0.0.0:10000
faraday-liionsden-azurite-1  | Azurite Blob service is successfully listening at http://0.0.0.0:10000
faraday-liionsden-azurite-1  | Azurite Queue service is starting at http://0.0.0.0:10001
faraday-liionsden-azurite-1  | Azurite Queue service is successfully listening at http://0.0.0.0:10001
faraday-liionsden-azurite-1  | Azurite Table service is starting at http://0.0.0.0:10002
faraday-liionsden-azurite-1  | Azurite Table service is successfully listening at http://0.0.0.0:10002
faraday-liionsden-db-1       | 2024-09-06 16:10:48.223 UTC [1] FATAL:  database files are incompatible with server
faraday-liionsden-db-1       | 2024-09-06 16:10:48.223 UTC [1] DETAIL:  The data directory was initialized by PostgreSQL version 13, which is not compatible with this version 15.8 (Debian 15.8-1.pgdg120+1).
faraday-liionsden-db-1 exited with code 1
dependency failed to start: container for service "db" exited (1)
```

It appears like this means this will not work with any existing databases. If there are no databases, then this isn't a problem.

**Comment by dalonsoa** [comment]
The whole of Liionsden resource group has been deleted, so there’s no database whatsoever.



----------------------------------
Dr Diego Alonso Álvarez (he/him)

Head of Research Software Engineering



[p.s If I send an email outside of your working hours please do not feel the urgency to respond until it is your working hours]



________________________________

**Review by AdrianDAlessandro — APPROVED** [review]
